File: lib/game-dictionaries.ts
/**
 * GameStringer Game Dictionaries Client
 * 
 * TypeScript client per la gestione dei dizionari di traduzione game-specific.
 * Integra con il backend Rust per import/export, ricerca, e applicazione a XUnity.
 */

import { invoke } from '@/lib/tauri-api';
import { clientLogger } from '@/lib/client-logger';

// Types
export interface DictionaryInfo {
  id: string;
  game_name: string;
  game_id?: string;
  source_language: string;
  target_language: string;
  entries_count: number;
  version: string;
  author?: string;
  description?: string;
  size_bytes: number;
  created_at: string;
  updated_at: string;
}

export interface GameDictionary {
  info: DictionaryInfo;
  translations: Record<string, string>;
}

export interface DictionaryResult {
  success: boolean;
  message: string;
  dictionary_id?: string;
  entries_loaded?: number;
}

export interface LocalDictionaryStatus {
  installed: boolean;
  dictionary_id?: string;
  entries_count: number;
  last_updated?: string;
  file_path?: string;
}

export interface DictionariesStats {
  total_dictionaries: number;
  total_entries: number;
  total_size_mb: number;
  by_language: Record<string, number>;
  dictionaries: DictionaryInfo[];
}

/**
 * Lista tutti i dizionari installati localmente
 */
export async function listInstalledDictionaries(): Promise<DictionaryInfo[]> {
  try {
    return await invoke<DictionaryInfo[]>('list_installed_dictionaries');
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore lista dizionari: ${String(error)}`, 'DICT');
    return [];
  }
}

/**
 * Carica un dizionario completo
 */
export async function loadDictionary(gameId: string, targetLang: string): Promise<GameDictionary | null> {
  try {
    return await invoke<GameDictionary>('load_dictionary', {
      gameId,
      targetLang,
    });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore caricamento dizionario: ${String(error)}`, 'DICT');
    return null;
  }
}

/**
 * Salva un dizionario
 */
export async function saveDictionary(dictionary: GameDictionary): Promise<DictionaryResult> {
  try {
    return await invoke<DictionaryResult>('save_dictionary', { dictionary });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore salvataggio dizionario: ${String(error)}`, 'DICT');
    return {
      success: false,
      message: `Errore: ${error}`,
    };
  }
}

/**
 * Importa un dizionario da file JSON (auto-rileva metadati da _meta)
 */
export async function importDictionaryAuto(filePath: string): Promise<DictionaryResult> {
  try {
    return await invoke<DictionaryResult>('import_dictionary_auto', {
      filePath: filePath,
    });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore importazione auto: ${String(error)}`, 'DICT');
    return {
      success: false,
      message: `Errore: ${error}`,
    };
  }
}

/**
 * Importa un dizionario da file JSON (parametri manuali)
 */
export async function importDictionaryFromFile(
  filePath: string,
  gameId: string,
  gameName: string,
  sourceLang: string,
  targetLang: string
): Promise<DictionaryResult> {
  try {
    return await invoke<DictionaryResult>('import_dictionary_from_file', {
      filePath,
      gameId,
      gameName,
      sourceLang,
      targetLang,
    });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore importazione: ${String(error)}`, 'DICT');
    return {
      success: false,
      message: `Errore: ${error}`,
    };
  }
}

/**
 * Esporta un dizionario in formato semplice (solo traduzioni)
 */
export async function exportDictionarySimple(
  gameId: string,
  targetLang: string,
  outputPath: string
): Promise<DictionaryResult> {
  try {
    return await invoke<DictionaryResult>('export_dictionary_simple', {
      gameId,
      targetLang,
      outputPath,
    });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore esportazione: ${String(error)}`, 'DICT');
    return {
      success: false,
      message: `Errore: ${error}`,
    };
  }
}

/**
 * Elimina un dizionario locale
 */
export async function deleteDictionary(gameId: string, targetLang: string): Promise<DictionaryResult> {
  try {
    return await invoke<DictionaryResult>('delete_dictionary', {
      gameId,
      targetLang,
    });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore eliminazione: ${String(error)}`, 'DICT');
    return {
      success: false,
      message: `Errore: ${error}`,
    };
  }
}

/**
 * Verifica se un dizionario è installato
 */
export async function getDictionaryStatus(gameId: string, targetLang: string): Promise<LocalDictionaryStatus> {
  try {
    return await invoke<LocalDictionaryStatus>('get_dictionary_status', {
      gameId,
      targetLang,
    });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore status dizionario: ${String(error)}`, 'DICT');
    return {
      installed: false,
      entries_count: 0,
    };
  }
}

/**
 * Cerca nel dizionario
 */
export async function searchInDictionary(
  gameId: string,
  targetLang: string,
  query: string,
  limit?: number
): Promise<[string, string][]> {
  try {
    return await invoke<[string, string][]>('search_in_dictionary', {
      gameId,
      targetLang,
      query,
      limit,
    });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore ricerca: ${String(error)}`, 'DICT');
    return [];
  }
}

/**
 * Aggiunge una traduzione al dizionario
 */
export async function addTranslationToDictionary(
  gameId: string,
  targetLang: string,
  original: string,
  translated: string
): Promise<DictionaryResult> {
  try {
    return await invoke<DictionaryResult>('add_translation_to_dictionary', {
      gameId,
      targetLang,
      original,
      translated,
    });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore aggiunta traduzione: ${String(error)}`, 'DICT');
    return {
      success: false,
      message: `Errore: ${error}`,
    };
  }
}

/**
 * Unisce due dizionari
 */
export async function mergeDictionaries(
  baseGameId: string,
  baseLang: string,
  mergeFilePath: string
): Promise<DictionaryResult> {
  try {
    return await invoke<DictionaryResult>('merge_dictionaries', {
      baseGameId,
      baseLang,
      mergeFilePath,
    });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore merge: ${String(error)}`, 'DICT');
    return {
      success: false,
      message: `Errore: ${error}`,
    };
  }
}

/**
 * Applica dizionario a XUnity AutoTranslator
 */
export async function applyDictionaryToXUnity(
  gameId: string,
  targetLang: string,
  gamePath: string
): Promise<DictionaryResult> {
  try {
    return await invoke<DictionaryResult>('apply_dictionary_to_xunity', {
      gameId,
      targetLang,
      gamePath,
    });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore applicazione XUnity: ${String(error)}`, 'DICT');
    return {
      success: false,
      message: `Errore: ${error}`,
    };
  }
}

/**
 * Estrae traduzioni da XUnity AutoTranslator
 */
export async function extractXUnityTranslations(
  gamePath: string,
  targetLang: string
): Promise<Record<string, string>> {
  try {
    return await invoke<Record<string, string>>('extract_xunity_translations', {
      gamePath,
      targetLang,
    });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore estrazione XUnity: ${String(error)}`, 'DICT');
    return {};
  }
}

/**
 * Importa traduzioni da XUnity nel dizionario GameStringer
 */
export async function importFromXUnity(
  gamePath: string,
  gameId: string,
  gameName: string,
  targetLang: string
): Promise<DictionaryResult> {
  try {
    return await invoke<DictionaryResult>('import_from_xunity', {
      gamePath,
      gameId,
      gameName,
      targetLang,
    });
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore import da XUnity: ${String(error)}`, 'DICT');
    return {
      success: false,
      message: `Errore: ${error}`,
    };
  }
}

/**
 * Ottieni statistiche aggregate sui dizionari
 */
export async function getDictionariesStats(): Promise<DictionariesStats | null> {
  try {
    return await invoke<DictionariesStats>('get_dictionaries_stats');
  } catch (error: unknown) {
    clientLogger.error(`[GameDictionaries] Errore statistiche: ${String(error)}`, 'DICT');
    return null;
  }
}

// Utility functions

/**
 * Formatta dimensione file
 */
export function formatFileSize(bytes: number): string {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(2))} ${sizes[i]}`;
}

/**
 * Formatta data
 */
export function formatDate(dateStr: string): string {
  try {
    const date = new Date(dateStr);
    return date.toLocaleDateString('it-IT', {
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
    });
  } catch {
    return dateStr;
  }
}

/**
 * Ottieni nome lingua da codice
 */
export function getLanguageName(code: string): string {
  const languages: Record<string, string> = {
    en: 'English',
    it: 'Italiano',
    de: 'Deutsch',
    fr: 'Français',
    es: 'Español',
    pt: 'Português',
    ru: 'Русский',
    ja: '日本語',
    ko: '한국어',
    zh: '中文',
    pl: 'Polski',
    nl: 'Nederlands',
    tr: 'Türkçe',
    ar: 'العربية',
  };
  return languages[code] || code.toUpperCase();
}

/**
 * Ottieni bandiera da codice lingua
 */
export function getLanguageFlag(code: string): string {
  const flags: Record<string, string> = {
    en: '🇬🇧',
    it: '🇮🇹',
    de: '🇩🇪',
    fr: '🇫🇷',
    es: '🇪🇸',
    pt: '🇵🇹',
    ru: '🇷🇺',
    ja: '🇯🇵',
    ko: '🇰🇷',
    zh: '🇨🇳',
    pl: '🇵🇱',
    nl: '🇳🇱',
    tr: '🇹🇷',
    ar: '🇸🇦',
  };
  return flags[code] || '🌐';
}

File: src-tauri/src/commands/game_dictionaries.rs
//! Game Dictionaries Management
//! 
//! Sistema per gestire dizionari di traduzione pre-caricati per giochi specifici.
//! Supporta download da repository, import/export locale, e integrazione con XUnity.

use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use serde::{Deserialize, Serialize};
use tauri::command;

/// Informazioni su un dizionario disponibile
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DictionaryInfo {
    pub id: String,
    pub game_name: String,
    pub game_id: Option<String>,
    pub source_language: String,
    pub target_language: String,
    pub entries_count: usize,
    pub version: String,
    pub author: Option<String>,
    pub description: Option<String>,
    pub size_bytes: u64,
    pub created_at: String,
    pub updated_at: String,
}

/// Dizionario di traduzione per un gioco
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GameDictionary {
    pub info: DictionaryInfo,
    pub translations: HashMap<String, String>,
}

/// Risultato operazione dizionario
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DictionaryResult {
    pub success: bool,
    pub message: String,
    pub dictionary_id: Option<String>,
    pub entries_loaded: Option<usize>,
}

/// Stato del dizionario locale
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LocalDictionaryStatus {
    pub installed: bool,
    pub dictionary_id: Option<String>,
    pub entries_count: usize,
    pub last_updated: Option<String>,
    pub file_path: Option<String>,
}

/// Ottiene il percorso della cartella dizionari
fn get_dictionaries_dir() -> Result<PathBuf, String> {
    let app_data = dirs::data_local_dir()
        .ok_or("Impossibile trovare cartella dati app")?;
    let dict_dir = app_data.join("GameStringer").join("dictionaries");
    
    fs::create_dir_all(&dict_dir)
        .map_err(|e| format!("Errore creazione cartella dizionari: {}", e))?;
    
    Ok(dict_dir)
}

/// Ottiene il percorso del dizionario per un gioco specifico
fn get_dictionary_path(game_id: &str, target_lang: &str) -> Result<PathBuf, String> {
    let dict_dir = get_dictionaries_dir()?;
    Ok(dict_dir.join(format!("{}_{}.json", game_id, target_lang)))
}

/// Lista tutti i dizionari installati localmente
#[command]
pub async fn list_installed_dictionaries() -> Result<Vec<DictionaryInfo>, String> {
    let dict_dir = get_dictionaries_dir()?;
    let mut dictionaries = Vec::new();
    
    let entries = fs::read_dir(&dict_dir)
        .map_err(|e| format!("Errore lettura cartella dizionari: {}", e))?;
    
    for entry in entries.flatten() {
        let path = entry.path();
        if path.extension().map(|e| e == "json").unwrap_or(false) {
            if let Ok(content) = fs::read_to_string(&path) {
                // Prova a leggere solo le info (formato compatto)
                if let Ok(dict) = serde_json::from_str::<GameDictionary>(&content) {
                    dictionaries.push(dict.info);
                } else if let Ok(info) = serde_json::from_str::<DictionaryInfo>(&content) {
                    dictionaries.push(info);
                }
            }
        }
    }
    
    // Ordina per nome gioco
    dictionaries.sort_by(|a, b| a.game_name.cmp(&b.game_name));
    
    Ok(dictionaries)
}

/// Carica un dizionario locale
#[command(rename_all = "camelCase")]
pub async fn load_dictionary(game_id: String, target_lang: String) -> Result<GameDictionary, String> {
    let path = get_dictionary_path(&game_id, &target_lang)?;
    
    if !path.exists() {
        return Err(format!("Dizionario non trovato per {} ({})", game_id, target_lang));
    }
    
    let content = fs::read_to_string(&path)
        .map_err(|e| format!("Errore lettura dizionario: {}", e))?;
    
    let dict: GameDictionary = serde_json::from_str(&content)
        .map_err(|e| format!("Errore parsing dizionario: {}", e))?;
    
    log::info!("📚 Caricato dizionario '{}' con {} voci", game_id, dict.translations.len());
    
    Ok(dict)
}

/// Salva un dizionario locale
#[command]
pub async fn save_dictionary(dictionary: GameDictionary) -> Result<DictionaryResult, String> {
    let path = get_dictionary_path(&dictionary.info.id, &dictionary.info.target_language)?;
    
    let content = serde_json::to_string_pretty(&dictionary)
        .map_err(|e| format!("Errore serializzazione dizionario: {}", e))?;
    
    fs::write(&path, content)
        .map_err(|e| format!("Errore salvataggio dizionario: {}", e))?;
    
    log::info!("💾 Salvato dizionario '{}' con {} voci", dictionary.info.id, dictionary.translations.len());
    
    Ok(DictionaryResult {
        success: true,
        message: format!("Dizionario salvato: {}", path.display()),
        dictionary_id: Some(dictionary.info.id),
        entries_loaded: Some(dictionary.translations.len()),
    })
}

/// Metadati opzionali nel file JSON
#[derive(Debug, Clone, Serialize, Deserialize)]
struct DictionaryMeta {
    game_id: String,
    game_name: String,
    source_language: Option<String>,
    target_language: Option<String>,
    version: Option<String>,
    author: Option<String>,
}

/// Formato JSON con metadati
#[derive(Debug, Clone, Serialize, Deserialize)]
struct DictionaryWithMeta {
    #[serde(rename = "_meta")]
    meta: DictionaryMeta,
    #[serde(flatten)]
    translations: HashMap<String, serde_json::Value>,
}

/// Importa un dizionario da file JSON (auto-rileva metadati da _meta)
#[command(rename_all = "camelCase")]
pub async fn import_dictionary_auto(
    file_path: String,
) -> Result<DictionaryResult, String> {
    let path = Path::new(&file_path);
    
    if !path.exists() {
        return Err("File non trovato".to_string());
    }
    
    let content = fs::read_to_string(path)
        .map_err(|e| format!("Errore lettura file: {}", e))?;
    
    // Prova a leggere con metadati
    if let Ok(with_meta) = serde_json::from_str::<DictionaryWithMeta>(&content) {
        let meta = with_meta.meta;
        
        // Filtra le traduzioni (escludi _meta)
        let translations: HashMap<String, String> = with_meta.translations
            .into_iter()
            .filter(|(k, _)| k != "_meta")
            .filter_map(|(k, v)| {
                if let serde_json::Value::String(s) = v {
                    Some((k, s))
                } else {
                    None
                }
            })
            .collect();
        
        let now = chrono::Utc::now().to_rfc3339();
        let file_size = fs::metadata(path).map(|m| m.len()).unwrap_or(0);
        
        let dictionary = GameDictionary {
            info: DictionaryInfo {
                id: meta.game_id.clone(),
                game_name: meta.game_name,
                game_id: Some(meta.game_id),
                source_language: meta.source_language.unwrap_or_else(|| "en".to_string()),
                target_language: meta.target_language.unwrap_or_else(|| "it".to_string()),
                entries_count: translations.len(),
                version: meta.version.unwrap_or_else(|| "1.0.0".to_string()),
                author: meta.author,
                description: Some(format!("Importato da {}", file_path)),
                size_bytes: file_size,
                created_at: now.clone(),
                updated_at: now,
            },
            translations,
        };
        
        log::info!("📚 Importato dizionario '{}' con {} voci (auto-detect)", dictionary.info.game_name, dictionary.translations.len());
        
        return save_dictionary(dictionary).await;
    }
    
    Err("File JSON deve contenere _meta con game_id e game_name".to_string())
}

/// Importa un dizionario da file JSON esterno (parametri manuali)
#[command(rename_all = "camelCase")]
pub async fn import_dictionary_from_file(
    file_path: String,
    game_id: String,
    game_name: String,
    source_lang: String,
    target_lang: String,
) -> Result<DictionaryResult, String> {
    let path = Path::new(&file_path);
    
    if !path.exists() {
        return Err("File non trovato".to_string());
    }
    
    let content = fs::read_to_string(path)
        .map_err(|e| format!("Errore lettura file: {}", e))?;
    
    // Supporta diversi formati
    let translations: HashMap<String, String> = if let Ok(with_meta) = serde_json::from_str::<DictionaryWithMeta>(&content) {
        // Ha metadati, filtra _meta
        with_meta.translations
            .into_iter()
            .filter(|(k, _)| k != "_meta")
            .filter_map(|(k, v)| {
                if let serde_json::Value::String(s) = v {
                    Some((k, s))
                } else {
                    None
                }
            })
            .collect()
    } else if let Ok(map) = serde_json::from_str(&content) {
        map
    } else if let Ok(dict) = serde_json::from_str::<GameDictionary>(&content) {
        dict.translations
    } else {
        return Err("Formato file non riconosciuto".to_string());
    };
    
    let now = chrono::Utc::now().to_rfc3339();
    let file_size = fs::metadata(path).map(|m| m.len()).unwrap_or(0);
    
    let dictionary = GameDictionary {
        info: DictionaryInfo {
            id: game_id.clone(),
            game_name,
            game_id: Some(game_id),
            source_language: source_lang,
            target_language: target_lang.clone(),
            entries_count: translations.len(),
            version: "1.0.0".to_string(),
            author: None,
            description: Some(format!("Importato da {}", file_path)),
            size_bytes: file_size,
            created_at: now.clone(),
            updated_at: now,
        },
        translations,
    };
    
    save_dictionary(dictionary).await
}

/// Esporta un dizionario in formato semplice (solo traduzioni)
#[command(rename_all = "camelCase")]
pub async fn export_dictionary_simple(
    game_id: String,
    target_lang: String,
    output_path: String,
) -> Result<DictionaryResult, String> {
    let dict = load_dictionary(game_id.clone(), target_lang).await?;
    
    let content = serde_json::to_string_pretty(&dict.translations)
        .map_err(|e| format!("Errore serializzazione: {}", e))?;
    
    fs::write(&output_path, content)
        .map_err(|e| format!("Errore scrittura file: {}", e))?;
    
    Ok(DictionaryResult {
        success: true,
        message: format!("Esportato in: {}", output_path),
        dictionary_id: Some(game_id),
        entries_loaded: Some(dict.translations.len()),
    })
}

/// Elimina un dizionario locale
#[command(rename_all = "camelCase")]
pub async fn delete_dictionary(game_id: String, target_lang: String) -> Result<DictionaryResult, String> {
    let path = get_dictionary_path(&game_id, &target_lang)?;
    
    if !path.exists() {
        return Err("Dizionario non trovato".to_string());
    }
    
    fs::remove_file(&path)
        .map_err(|e| format!("Errore eliminazione dizionario: {}", e))?;
    
    log::info!("🗑️ Eliminato dizionario '{}_{}'", game_id, target_lang);
    
    Ok(DictionaryResult {
        success: true,
        message: format!("Dizionario {} eliminato", game_id),
        dictionary_id: Some(game_id),
        entries_loaded: None,
    })
}

/// Verifica se un dizionario è installato per un gioco
#[command(rename_all = "camelCase")]
pub async fn get_dictionary_status(game_id: String, target_lang: String) -> Result<LocalDictionaryStatus, String> {
    let path = get_dictionary_path(&game_id, &target_lang)?;
    
    if !path.exists() {
        return Ok(LocalDictionaryStatus {
            installed: false,
            dictionary_id: None,
            entries_count: 0,
            last_updated: None,
            file_path: None,
        });
    }
    
    let content = fs::read_to_string(&path)
        .map_err(|e| format!("Errore lettura dizionario: {}", e))?;
    
    let dict: GameDictionary = serde_json::from_str(&content)
        .map_err(|e| format!("Errore parsing dizionario: {}", e))?;
    
    Ok(LocalDictionaryStatus {
        installed: true,
        dictionary_id: Some(dict.info.id),
        entries_count: dict.translations.len(),
        last_updated: Some(dict.info.updated_at),
        file_path: Some(path.to_string_lossy().to_string()),
    })
}

/// Cerca nel dizionario
#[command(rename_all = "camelCase")]
pub async fn search_in_dictionary(
    game_id: String,
    target_lang: String,
    query: String,
    limit: Option<usize>,
) -> Result<Vec<(String, String)>, String> {
    let dict = load_dictionary(game_id, target_lang).await?;
    let limit = limit.unwrap_or(50);
    let query_lower = query.to_lowercase();
    
    let results: Vec<(String, String)> = dict.translations
        .iter()
        .filter(|(k, v)| {
            k.to_lowercase().contains(&query_lower) || 
            v.to_lowercase().contains(&query_lower)
        })
        .take(limit)
        .map(|(k, v)| (k.clone(), v.clone()))
        .collect();
    
    Ok(results)
}

/// Aggiunge o aggiorna una traduzione nel dizionario
#[command(rename_all = "camelCase")]
pub async fn add_translation_to_dictionary(
    game_id: String,
    target_lang: String,
    original: String,
    translated: String,
) -> Result<DictionaryResult, String> {
    let path = get_dictionary_path(&game_id, &target_lang)?;
    
    let mut dict = if path.exists() {
        load_dictionary(game_id.clone(), target_lang.clone()).await?
    } else {
        // Crea nuovo dizionario
        let now = chrono::Utc::now().to_rfc3339();
        GameDictionary {
            info: DictionaryInfo {
                id: game_id.clone(),
                game_name: game_id.clone(),
                game_id: Some(game_id.clone()),
                source_language: "en".to_string(),
                target_language: target_lang.clone(),
                entries_count: 0,
                version: "1.0.0".to_string(),
                author: None,
                description: None,
                size_bytes: 0,
                created_at: now.clone(),
                updated_at: now,
            },
            translations: HashMap::new(),
        }
    };
    
    // Aggiorna traduzione
    dict.translations.insert(original, translated);
    dict.info.entries_count = dict.translations.len();
    dict.info.updated_at = chrono::Utc::now().to_rfc3339();
    
    save_dictionary(dict).await
}

/// Unisce due dizionari (il secondo sovrascrive in caso di conflitto)
#[command(rename_all = "camelCase")]
pub async fn merge_dictionaries(
    base_game_id: String,
    base_lang: String,
    merge_file_path: String,
) -> Result<DictionaryResult, String> {
    let mut base_dict = load_dictionary(base_game_id.clone(), base_lang.clone()).await?;
    
    let merge_content = fs::read_to_string(&merge_file_path)
        .map_err(|e| format!("Errore lettura file da unire: {}", e))?;
    
    let merge_translations: HashMap<String, String> = serde_json::from_str(&merge_content)
        .map_err(|e| format!("Errore parsing file da unire: {}", e))?;
    
    let added = merge_translations.len();
    
    // Unisci (sovrascrive duplicati)
    for (k, v) in merge_translations {
        base_dict.translations.insert(k, v);
    }
    
    base_dict.info.entries_count = base_dict.translations.len();
    base_dict.info.updated_at = chrono::Utc::now().to_rfc3339();
    
    let result = save_dictionary(base_dict).await?;
    
    Ok(DictionaryResult {
        success: true,
        message: format!("Uniti {} voci, totale: {}", added, result.entries_loaded.unwrap_or(0)),
        dictionary_id: result.dictionary_id,
        entries_loaded: result.entries_loaded,
    })
}

/// Applica dizionario a XUnity AutoTranslator (copia in cartella gioco)
#[command(rename_all = "camelCase")]
pub async fn apply_dictionary_to_xunity(
    game_id: String,
    target_lang: String,
    game_path: String,
) -> Result<DictionaryResult, String> {
    let dict = load_dictionary(game_id.clone(), target_lang.clone()).await?;
    
    // Percorso XUnity translations
    let xunity_dir = Path::new(&game_path)
        .join("BepInEx")
        .join("Translation")
        .join(&target_lang)
        .join("Text");
    
    fs::create_dir_all(&xunity_dir)
        .map_err(|e| format!("Errore creazione cartella XUnity: {}", e))?;
    
    // XUnity usa formato: originale=traduzione (una per riga)
    let mut content = String::new();
    for (original, translated) in &dict.translations {
        // Escape caratteri speciali
        let orig_escaped = original.replace("=", "\\=").replace("\n", "\\n");
        let trans_escaped = translated.replace("=", "\\=").replace("\n", "\\n");
        content.push_str(&format!("{}={}\n", orig_escaped, trans_escaped));
    }
    
    let output_file = xunity_dir.join("_GameStringer.txt");
    fs::write(&output_file, content)
        .map_err(|e| format!("Errore scrittura file XUnity: {}", e))?;
    
    log::info!("✅ Applicato dizionario a XUnity: {}", output_file.display());
    
    Ok(DictionaryResult {
        success: true,
        message: format!("Dizionario applicato a XUnity in: {}", output_file.display()),
        dictionary_id: Some(game_id),
        entries_loaded: Some(dict.translations.len()),
    })
}

/// Estrae stringhe da XUnity (file _AutoGeneratedTranslations.txt)
#[command(rename_all = "camelCase")]
pub async fn extract_xunity_translations(
    game_path: String,
    target_lang: String,
) -> Result<HashMap<String, String>, String> {
    let xunity_file = Path::new(&game_path)
        .join("BepInEx")
        .join("Translation")
        .join(&target_lang)
        .join("Text")
        .join("_AutoGeneratedTranslations.txt");
    
    if !xunity_file.exists() {
        return Err("File traduzioni XUnity non trovato. Avvia il gioco almeno una volta con BepInEx.".to_string());
    }
    
    let content = fs::read_to_string(&xunity_file)
        .map_err(|e| format!("Errore lettura file XUnity: {}", e))?;
    
    let mut translations = HashMap::new();
    
    for line in content.lines() {
        if line.starts_with("//") || line.trim().is_empty() {
            continue;
        }
        
        if let Some(idx) = line.find('=') {
            let original = line[..idx].replace("\\=", "=").replace("\\n", "\n");
            let translated = line[idx + 1..].replace("\\=", "=").replace("\\n", "\n");
            
            if !original.is_empty() && !translated.is_empty() {
                translations.insert(original, translated);
            }
        }
    }
    
    log::info!("📤 Estratte {} traduzioni da XUnity", translations.len());
    
    Ok(translations)
}

/// Importa traduzioni da XUnity nel dizionario GameStringer
#[command(rename_all = "camelCase")]
pub async fn import_from_xunity(
    game_path: String,
    game_id: String,
    game_name: String,
    target_lang: String,
) -> Result<DictionaryResult, String> {
    let translations = extract_xunity_translations(game_path, target_lang.clone()).await?;
    
    if translations.is_empty() {
        return Err("Nessuna traduzione trovata in XUnity".to_string());
    }
    
    let now = chrono::Utc::now().to_rfc3339();
    
    let dictionary = GameDictionary {
        info: DictionaryInfo {
            id: game_id.clone(),
            game_name,
            game_id: Some(game_id),
            source_language: "en".to_string(),
            target_language: target_lang,
            entries_count: translations.len(),
            version: "1.0.0".to_string(),
            author: Some("XUnity AutoTranslator".to_string()),
            description: Some("Importato da traduzioni automatiche XUnity".to_string()),
            size_bytes: 0,
            created_at: now.clone(),
            updated_at: now,
        },
        translations,
    };
    
    save_dictionary(dictionary).await
}

/// Statistiche aggregate sui dizionari
#[command]
pub async fn get_dictionaries_stats() -> Result<serde_json::Value, String> {
    let dictionaries = list_installed_dictionaries().await?;
    
    let total_entries: usize = dictionaries.iter().map(|d| d.entries_count).sum();
    let total_size: u64 = dictionaries.iter().map(|d| d.size_bytes).sum();
    
    let mut by_language: HashMap<String, usize> = HashMap::new();
    for dict in &dictionaries {
        *by_language.entry(dict.target_language.clone()).or_insert(0) += 1;
    }
    
    Ok(serde_json::json!({
        "total_dictionaries": dictionaries.len(),
        "total_entries": total_entries,
        "total_size_mb": total_size as f64 / 1024.0 / 1024.0,
        "by_language": by_language,
        "dictionaries": dictionaries,
    }))
}
